From 099f122a42098f5150f221b6d912a5d262b74643 Mon Sep 17 00:00:00 2001
From: wfeng6 <wei.feng@intel.com>
Date: Wed, 22 May 2013 18:40:05 +0800
Subject: [PORT FROM MAIN]Fix the video editor video decoder render problem

BZ: 108883 108896

In some situations the video editor decoder part will provide a useless
frame when generating the first frame of the transition clip. The patch
fix this problem.

Change-Id: I5ecdb36498001baa681f0e9a8f37a095b917f23b
Orig-Change-Id: If8c018e8c2626dc6f12816f64d4d0671f1d3d522
Signed-off-by: wfeng6 <wei.feng@intel.com>
Signed-off-by: ywan171 <yi.a.wang@intel.com>
---
 .../src/VideoEditorVideoDecoder.cpp                |   31 ++++++++++++++++++-
 1 files changed, 29 insertions(+), 2 deletions(-)
 mode change 100755 => 100644 libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp

diff --git a/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp b/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp
old mode 100755
new mode 100644
index 1845480..417451e
--- a/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp
+++ b/libvideoeditor/vss/stagefrightshells/src/VideoEditorVideoDecoder.cpp
@@ -1587,6 +1587,10 @@ M4OSA_ERR VideoEditorVideoDecoder_render(M4OSA_Context context,
     M4_MediaTime candidateTimeStamp = -1;
     M4OSA_Bool bFound = M4OSA_FALSE;
 
+    M4_MediaTime minTimeStampDifference = 100000;
+    M4_MediaTime mTimeStampDifference;
+    M4OSA_UInt32 matchBufferIdx;
+
     ALOGV("VideoEditorVideoDecoder_render begin");
     // Input parameters check
     VIDEOEDITOR_CHECK(M4OSA_NULL != context, M4ERR_PARAMETER);
@@ -1629,8 +1633,31 @@ M4OSA_ERR VideoEditorVideoDecoder_render(M4OSA_Context context,
         }
     }
     if (M4OSA_FALSE == bFound) {
-        err = M4WAR_VIDEORENDERER_NO_NEW_FRAME;
-        goto cleanUp;
+            // Sometimes the nearest cts is larger than *pTime
+        for (i=0; i < pDecShellContext->m_pDecBufferPool->NB; i++) {
+            pTmpVIDEOEDITORBuffer = &pDecShellContext->m_pDecBufferPool->pNXPBuffer[i];
+            if (pTmpVIDEOEDITORBuffer->state == VIDEOEDITOR_BUFFER_kFilled) {
+                mTimeStampDifference = abs(pTmpVIDEOEDITORBuffer->buffCTS - *pTime);
+                if (minTimeStampDifference > mTimeStampDifference)
+                {
+                    minTimeStampDifference = mTimeStampDifference;
+                    matchBufferIdx = i;
+                    bFound = M4OSA_TRUE;
+                    ALOGV("mTimeStampDifference = %lf, matchBufferIdx = %d",minTimeStampDifference,matchBufferIdx);
+                }
+            }
+        }
+        if (bFound) {
+            pRenderVIDEOEDITORBuffer = &pDecShellContext->m_pDecBufferPool->pNXPBuffer[matchBufferIdx];
+            candidateTimeStamp = pRenderVIDEOEDITORBuffer->buffCTS;
+            ALOGV("VideoDecoder_render_NV12: found a buffer with nearest timestamp = %lf",
+                    candidateTimeStamp);
+        }
+        else
+        {
+            err = M4WAR_VIDEORENDERER_NO_NEW_FRAME;
+            goto cleanUp;
+        }
     }
 
     ALOGV("VideoEditorVideoDecoder_render 3 ouput %d %d %d %d",
-- 
1.7.4.1

